### ISSUE TITLE
pip broken with strict version check

### ISSUE THREAD

**Issue by senges** [NixOS/nixpkgs#198059]
### Describe the bug
I'm using python package manager `pip` inside a `nixos/nix:2.11.1` docker container.
Every time a new pip version is released, pip version check breaks the whole installation process :

```
...
Message: '[present-any] %s'
Arguments: (UpgradePrompt(old='22.2.2', new='22.3'),)  
```

<details>
  <summary>Full stack trace</summary>

**Example installing `rich` package**

  ```
$  pip3 install --user rich

Collecting rich
  Using cached rich-12.6.0-py3-none-any.whl (237 kB)
Collecting pygments<3.0.0,>=2.6.0
  Using cached Pygments-2.13.0-py3-none-any.whl (1.1 MB)
Collecting commonmark<0.10.0,>=0.9.0
  Using cached commonmark-0.9.1-py2.py3-none-any.whl (51 kB)
Installing collected packages: commonmark, pygments, rich
Successfully installed commonmark-0.9.1 pygments-2.13.0 rich-12.6.0
--- Logging error ---
Traceback (most recent call last):
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/utils/logging.py", line 177, in emit
    self.console.print(renderable, overflow="ignore", crop=False, style=style)
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_vendor/rich/console.py", line 1673, in print
    extend(render(renderable, render_options))
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_vendor/rich/console.py", line 1305, in render
    for render_output in iter_render:
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/utils/logging.py", line 134, in __rich_console__
    for line in lines:
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_vendor/rich/segment.py", line 249, in split_lines
    for segment in segments:
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_vendor/rich/console.py", line 1283, in render
    renderable = rich_cast(renderable)
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_vendor/rich/protocol.py", line 36, in rich_cast
    renderable = cast_method()
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/self_outdated_check.py", line 130, in __rich__
    pip_cmd = get_best_invocation_for_this_pip()
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/utils/entrypoints.py", line 58, in get_best_invocation_for_this_pip
    if found_executable and os.path.samefile(
  File "/nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/python3.10/genericpath.py", line 101, in samefile
    s2 = os.stat(f2)
FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/bin/pip'
Call stack:
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/bin/.pip3-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/cli/main.py", line 70, in main
    return command.main(cmd_args)
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/cli/base_command.py", line 101, in main
    return self._main(args)
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/cli/base_command.py", line 223, in _main
    self.handle_pip_version_check(options)
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/cli/req_command.py", line 190, in handle_pip_version_check
    pip_self_version_check(session, options)
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/self_outdated_check.py", line 236, in pip_self_version_check
    logger.warning("[present-rich] %s", upgrade_prompt)
  File "/nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/python3.10/logging/__init__.py", line 1489, in warning
    self._log(WARNING, msg, args, **kwargs)
  File "/nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/python3.10/logging/__init__.py", line 1624, in _log
    self.handle(record)
  File "/nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/python3.10/logging/__init__.py", line 1634, in handle
    self.callHandlers(record)
  File "/nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/python3.10/logging/__init__.py", line 1696, in callHandlers
    hdlr.handle(record)
  File "/nix/store/9srs642k875z3qdk8glapjycncf2pa51-python3-3.10.7/lib/python3.10/logging/__init__.py", line 968, in handle
    self.emit(record)
  File "/nix/store/0vixrqsg7j2m1hjyd2m7j7k58bcd8bqi-python3.10-pip-22.2.2/lib/python3.10/site-packages/pip/_internal/utils/logging.py", line 179, in emit
    self.handleError(record)
Message: '[present-rich] %s'
Arguments: (UpgradePrompt(old='22.2.2', new='22.3'),)     
  ```
  
</details>

### Temp fix

Running pip install with `--disable-pip-version-check` flag disables the version check and pip works as expected.

### Steps To Reproduce
Tools are installed with `nix-env` as :

```
export NIXPKGS_ALLOW_UNFREE=1
nix-channel --update
nix-env -p /out/profile -iA nixpkgs.python310
nix-env -p /out/profile -iA nixpkgs.python310Packages.pip
```

### Expected behavior
pip might raise a warning but should probably not completely fail ?

Any known solution to this kind of problem ? :thinking:

**Comment by senges** [comment]
Indeed this is clearly a design mismatch between pip's way of locating pip executable and Nix way of installing tools.

I was digging a bit and I found an interesting issue which looks kinda related to my problem : https://github.com/pypa/pip/issues/5599

However, I will follow your suggestion and see how this issue can be addressed in pip implementation itself.
Will keep this thread up to date, might interest someone in the future as pip is a quite useful tool.

Thanks for you answer

**Comment by senges** [comment]
As discussed on the `pypa/pip` repo mentioned above, this issue was just fixed a few weeks ago with the release of pip `22.3`.

To make sure pip version check error was also fixed on NixOS by the new release, I applied manually the patch by replacing `entrypoint.py` on my local `22.2.2` pip installation. Worked like a charm, we can resolve and close this issue.

<details>
  <summary>Patched pip install</summary>
  
  ```
pip install --user rich
Collecting rich
  Using cached rich-12.6.0-py3-none-any.whl (237 kB)
Collecting pygments<3.0.0,>=2.6.0
  Using cached Pygments-2.13.0-py3-none-any.whl (1.1 MB)
Collecting commonmark<0.10.0,>=0.9.0
  Using cached commonmark-0.9.1-py2.py3-none-any.whl (51 kB)
Installing collected packages: commonmark, pygments, rich
Successfully installed commonmark-0.9.1 pygments-2.13.0 rich-12.6.0

[notice] A new release of pip available: 22.2.2 -> 22.3
[notice] To update, run: python3.10 -m pip install --upgrade pip
  ```
</details>

**Comment by senges** [comment]
Hello @mweinelt I've just seen you were the last person to provide `python310Packages.pip` version update :wave: 

> TLDR; actual version of pip `22.2.2` is broken on Nix due to a preliminary version check. pip is looking for its executable in the wrong directory. New pip version `23.0` is fixing the bug.

It's a quite annoying issue, and I suppose I'm not the only person using pip in the Nix world.
I was considering opening a simple PR in`nixos-unstable` branch to update `version` and `sha256` in [python pip module definition](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/python-modules/pip/default.nix).

But I'm still quite new to the Nix world and I'm not sure how this situation should be handled. I'm far from having the technical abilities to do extensive testings and make sure this new version isn't breaking anything or introducing new bugs.

Deeply sorry for the unsolicited ping, but as pip is a quite important piece of software I thought I might bring this bug to your attention :pray:

**Comment by FRidh** [comment]
I've added a pip update to https://github.com/NixOS/nixpkgs/pull/198024.

**Comment by senges** [comment]
Nice thanks again !

**Comment by avanc** [comment]
Is there a workaround until the pull request is merged?
As buildPythonApplication somehow depends on pip, I have the same problem with my derivation:
```
{ lib, python3Packages }:
with python3Packages;
buildPythonApplication {
  pname = "mqtt-tools-timestamp";
  version = "0.1";

  propagatedBuildInputs = [ asyncio-mqtt ];

  src = ./.;
}
```

nix-shell returns the same error regarding pip.

**Comment by senges** [comment]
[Here are](https://stackoverflow.com/a/46288945) many ways of ignoring version check with pip. Is it what you are looking for ?

**Comment by avanc** [comment]
@senges Thanks for the hints. But I have to admit that I even don't know why the above expression with nox-shell calls pip at all.

**Comment by senges** [comment]
Well tbh my understanding of Nix isn't good enough to offer anything else :smile:
